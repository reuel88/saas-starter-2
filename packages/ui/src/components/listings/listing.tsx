import { FC, ReactNode } from "react";
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "../card";
import { Button } from "../button";

export type ListingProps = {
  title: string;
  price: number;
  location?: string;
  imageSrc?: string;
  description: ReactNode;
};
export const Listing: FC<ListingProps> = ({
  title,
  price,
  location,
  imageSrc,
  description,
  ...rest
}) => {
  return (
    <Card as="article" className="flex flex-col" {...rest}>
      <CardHeader as="header">
        <CardTitle className="line-clamp-1">{title}</CardTitle>
        <div className="flex items-center justify-between gap-4">
          <p className="text-lg font-extrabold text-hue-color">
            {new Intl.NumberFormat("en-AU", {
              style: "currency",
              currency: "AUD",
              minimumFractionDigits: 0,
            }).format(price)}
          </p>
          {location && (
            <p className="text-sm text-muted-foreground">{location}</p>
          )}
        </div>
      </CardHeader>

      <CardContent className="space-y-6">
        {imageSrc && (
          <img
            src={imageSrc}
            alt={title}
            className="aspect-video object-cover"
          />
        )}

        <CardDescription className="line-clamp-3">
          {description}
        </CardDescription>
      </CardContent>

      <CardFooter as="footer" className="mt-auto flex gap-4">
        <Button className="min-w-20" type="button">
          View
        </Button>
        <Button variant="outline" className="min-w-20" type="button">
          Reply
        </Button>
      </CardFooter>
    </Card>
  );
};
