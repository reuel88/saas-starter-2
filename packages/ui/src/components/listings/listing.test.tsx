import { render, screen } from "@testing-library/react";
import { Listing } from "./listing";

describe("Listing", () => {
  const renderListing = (props: any) => render(<Listing {...props} />);

  const testProps = {
    title: "Test Listing",
    price: 100,
    location: "Test Location",
    description: "Test Description",
  };

  it("should display the listing details", () => {
    // Arrange

    // Act
    const { baseElement } = renderListing(testProps);

    // Assert
    Object.values(testProps).forEach((value) => {
      if (typeof value === "string") {
        expect(screen.getByText(value.toString())).toBeInTheDocument();
      } else if (value === 100) {
        // This is the price
        expect(screen.getByText("$100")).toBeInTheDocument();
      }
    });
    expect(baseElement).toMatchSnapshot();
  });

  it("should format the price as an Australian dollar currency", () => {
    // Arrange

    // Act
    const { baseElement } = renderListing({ ...testProps, price: 1000 });

    // Assert
    expect(screen.getByText("$1,000")).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should display the image with the cover-fill and aspect-video classes", () => {
    // Arrange

    // Act
    const { baseElement } = renderListing({
      ...testProps,
      imageSrc: "test.jpg",
    });
    const image = screen.getByAltText(testProps.title);

    // Assert
    expect(image).toBeInTheDocument();
    expect(image).toHaveClass("object-cover");
    expect(image).toHaveClass("aspect-video");
    expect(baseElement).toMatchSnapshot();
  });
});
