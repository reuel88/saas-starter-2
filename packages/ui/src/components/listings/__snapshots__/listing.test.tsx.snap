// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Listing > should display the image with the cover-fill and aspect-video classes 1`] = `
<body>
  <div>
    <article
      class="rounded-lg border bg-card text-card-foreground shadow-sm flex flex-col"
    >
      <header
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="text-2xl font-semibold leading-none tracking-tight line-clamp-1"
        >
          Test Listing
        </h3>
        <div
          class="flex items-center justify-between gap-4"
        >
          <p
            class="text-lg font-extrabold text-hue-color"
          >
            $100
          </p>
          <p
            class="text-sm text-muted-foreground"
          >
            Test Location
          </p>
        </div>
      </header>
      <div
        class="p-6 pt-0 space-y-6"
      >
        <img
          alt="Test Listing"
          class="aspect-video object-cover"
          src="test.jpg"
        />
        <p
          class="text-sm text-muted-foreground line-clamp-3"
        >
          Test Description
        </p>
      </div>
      <footer
        class="items-center p-6 pt-0 mt-auto flex gap-4"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 min-w-20"
          type="button"
        >
          View
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 min-w-20"
          type="button"
        >
          Reply
        </button>
      </footer>
    </article>
  </div>
</body>
`;

exports[`Listing > should display the listing details 1`] = `
<body>
  <div>
    <article
      class="rounded-lg border bg-card text-card-foreground shadow-sm flex flex-col"
    >
      <header
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="text-2xl font-semibold leading-none tracking-tight line-clamp-1"
        >
          Test Listing
        </h3>
        <div
          class="flex items-center justify-between gap-4"
        >
          <p
            class="text-lg font-extrabold text-hue-color"
          >
            $100
          </p>
          <p
            class="text-sm text-muted-foreground"
          >
            Test Location
          </p>
        </div>
      </header>
      <div
        class="p-6 pt-0 space-y-6"
      >
        <p
          class="text-sm text-muted-foreground line-clamp-3"
        >
          Test Description
        </p>
      </div>
      <footer
        class="items-center p-6 pt-0 mt-auto flex gap-4"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 min-w-20"
          type="button"
        >
          View
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 min-w-20"
          type="button"
        >
          Reply
        </button>
      </footer>
    </article>
  </div>
</body>
`;

exports[`Listing > should format the price as an Australian dollar currency 1`] = `
<body>
  <div>
    <article
      class="rounded-lg border bg-card text-card-foreground shadow-sm flex flex-col"
    >
      <header
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="text-2xl font-semibold leading-none tracking-tight line-clamp-1"
        >
          Test Listing
        </h3>
        <div
          class="flex items-center justify-between gap-4"
        >
          <p
            class="text-lg font-extrabold text-hue-color"
          >
            $1,000
          </p>
          <p
            class="text-sm text-muted-foreground"
          >
            Test Location
          </p>
        </div>
      </header>
      <div
        class="p-6 pt-0 space-y-6"
      >
        <p
          class="text-sm text-muted-foreground line-clamp-3"
        >
          Test Description
        </p>
      </div>
      <footer
        class="items-center p-6 pt-0 mt-auto flex gap-4"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 min-w-20"
          type="button"
        >
          View
        </button>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 min-w-20"
          type="button"
        >
          Reply
        </button>
      </footer>
    </article>
  </div>
</body>
`;
